### PR TITLE
[GitHub Actions] Cron job for resolving pending metadata

### PR DESCRIPTION
**Pull request by KyleJu**
Fix for https://github.com/web-platform-tests/wpt.fyi/issues/2362. 

Schedule a cron job that resolves all pending metadata from /api/metadata/pending on a daily basis. This crob job will resolve all the pending metadata in a single PR. As a result, we don't have to resolve the cases in https://github.com/web-platform-tests/wpt.fyi/issues/2362 manually.

The `main.go` script is a rework of https://github.com/web-platform-tests/wpt-metadata/pull/473

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+git checkout -- || true
+
+# Resolve pending metadata
+./main
+
+# Create a PR
+git checkout -b resolve-daily-conflict
+git add .
+git commit -m 'Resolve daily merge conflicts'
+git push
```
Comment (KyleJu) [inline build.sh:19]: @foolip Do you know how to set credentials in a cron job? I want to create a branch and a PR automatically here?

**Comment by KyleJu** [comment]
@foolip I need a little help with `build.sh`. Its ready for a review otherwise

File: .github/workflows/resolve_merge_conflict.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+on:
+  schedule:
+    - cron: '0 0 * * *'
+jobs:
+  resolve-merge-conflicts:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: actions/setup-go@v2
```
Comment (KyleJu) [inline .github/workflows/resolve_merge_conflict.yml:11]: If I only run `./main` in `./build.sh`, I won't need golang environment?

File: build.sh (hunk 1 above)
Comment (KyleJu) [reply build.sh:19]: maybe it should create a PR, stop auto-merge and assign a reviewer?

Comment (foolip) [reply build.sh:19]: https://github.com/Ecosystem-Infra/wpt-results/blob/main/.github/workflows/results.yml has an example of how to use a token to push to a branch.

But since the intention is to create a PR, I would recommend using `peter-evans/create-pull-request` which handles lots of the boring details. https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/interfaces.yml has an example of that.

File: .github/workflows/resolve_merge_conflict.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+name: 'Resolve Merge Conflicts'
+on:
+  schedule:
+    - cron: '0 0 * * *'
+jobs:
+  resolve-merge-conflicts:
+    runs-on: ubuntu-latest
```
Comment (foolip) [inline .github/workflows/resolve_merge_conflict.yml:7]:
```suggestion
    runs-on: ubuntu-20.04
```

I'd recommend pinning the version of Ubuntu, so that things don't suddenly start failing. In https://github.com/web-platform-tests/wpt-metadata/pull/1155 I pinned everything to ubuntu-20.04, so suggesting the same here.

File: .github/workflows/resolve_merge_conflict.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: actions/setup-go@v2
+      with:
+        go-version: 1.13.x
+
+    - name: Build app
```
Comment (foolip) [inline .github/workflows/resolve_merge_conflict.yml:15]:
```suggestion
    - name: Run build.sh
```

No name is really needed, so just omitting it would also be OK.

File: build.sh [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+set -x
+set -o errexit
+set -o pipefail
+
+# Get a clean checkout
```
Comment (foolip) [inline build.sh:6]: We should already have a clean checkout when running in CI. If you rename this script to `resolve-conflicts.sh` or similar, it should be more clear it's not supposed to be run locally as a matter of course, and this can be removed.

File: build.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+git fetch origin || true
+git checkout origin/master || true
+git reset HEAD || true
+git checkout -- || true
+
+# Resolve pending metadata
+./main
```
Comment (foolip) [inline build.sh:13]: I see `./main` is checked into the repo. It should be built in the CI workflow instead, right?

**Comment by foolip** [comment]
I'd also suggest adding workflow_dispatch to make it easier to trigger manually once landed, see https://github.com/web-platform-tests/wpt/blob/master/.github/workflows/interfaces.yml#L6

File: build.sh (hunk 1 above)
Comment (KyleJu) [reply build.sh:19]: with `peter-evans/create-pull-request`, I can delete build.sh

**Comment by KyleJu** [comment]
@foolip Had a test run [here](https://github.com/web-platform-tests/wpt-metadata/actions/workflows/resolve_merge_conflict.yml). Ready for another look

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -13,4 +13,5 @@ jobs:
         go-version: 1.13.x
     - uses: actions/checkout@v2
     - run: go get -v -t -d ./...
-    - run: go test -v ./...
+    - run: go test -v parse_test.go
```
Comment (foolip) [inline .github/workflows/test.yml:16]: `parse_test.go` and `test_paths_test.go` aren't added in this PR, is this change related?

It does seem like some tests for the new code would be useful however, so maybe add a `test_main.go`?

File: main.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	for index, existingLink := range metadata.Links {
+		if !link.Product.MatchesProductSpec(existingLink.Product) {
+			continue
+		}
+
+		// Check whether we are already listed in this entry.
+		for _, result := range existingLink.Results {
+			if result.TestPath == testpath {
+				fmt.Println("EXISTING-LINK: for test", testName, "with url", link.URL)
+				return
```
Comment (foolip) [inline main.go:97]: Returning in the middle of this for loop, doesn't that mean that another link after this one could be new and is now going to be silently dropped on the floor? If this is a bug, can you add a test for this case?

To make it more obvious what's going on, would it be possible to structure this as looping through the smallest "atoms" of metadata we could add, and for each one check if the equivalent is already there and otherwise add it? Or would that require looping through all existing metadata for each addition because of how the data happens to be structured?

File: main.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
```
Comment (foolip) [inline main.go:57]: Oh neat, https://golang.org/pkg/fmt/#Println adds spaces. (Was about to suggest adding them, like other languages I know would require.)

File: main.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// First check whether we have an existing crbug link in this META.yml; if
```
Comment (foolip) [inline main.go:84]:
```suggestion
	// First check whether we have an existing link in this META.yml; if
```

(Might not be a crbug.com link.)

File: main.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+		if existingLink.URL == link.URL {
+			fmt.Println("APPEND-LINK: for test", testName, "with url", link.URL)
+			metadata.Links[index].Results = append(existingLink.Results, shared.MetadataTestResult{TestPath: testpath})
+			hasAdded = true
+			break
+		}
+	}
+
+	// Otherwise, this is a brand new crbug entry in the META.yml, so add a new
```
Comment (foolip) [inline main.go:109]:
```suggestion
	// Otherwise, this is a brand new entry in the META.yml, so add a new
```

Just to make sure I understand it, an "entry" is each unique a product+url an entry, and tests are part of that entry, right?

File: main.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		}
+		metadata.Links = append(metadata.Links, newLink)
+	}
+
+	writeToFile(metadata, f)
+}
+
+// createDirIfNeeded is a helper function to create a directory if it doesn't
+// already exist. It handles recursive directory creation.
+func createDirIfNeeded(dir string) {
```
Comment (foolip) [inline main.go:128]: It looks like just calling [`os.MkdirAll`](https://golang.org/pkg/os/#MkdirAll) would work here, per the documentation it does nothing if the directory already exists.

File: main.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	_, err := os.Stat(dir)
+	if !os.IsNotExist(err) {
+		return
+	}
+	os.MkdirAll(dir, os.ModePerm)
+}
+
+// writeToFile writes out a wpt.fyi shared.Metadata entry to a given file handle.
+//
+// TODO: This seems to cause a lot of existing metadata files to be completely
```
Comment (foolip) [inline main.go:138]: Could it be https://github.com/web-platform-tests/wpt.fyi/issues/1957?

This TODO seems serious enough that it should be resolved before merging, otherwise it's hard to vet that no unexpected changes are happening.

Maybe all updates to META.yml files should go through a shared code path, so that different code paths aren't flipping the formatting back and forth between different indentations and stuff?

File: .github/workflows/resolve_merge_conflict.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        go-version: 1.13.x
+
+    - name: build main
+      run: go build main.go
+
+    - name: run main
+      run: ./main
+
+    - name: Create pull request
+      uses: peter-evans/create-pull-request@v3
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply .github/workflows/resolve_merge_conflict.yml:23]: SG. I added `author: wpt-pr-bot <wpt-pr-bot@users.noreply.github.com>`

File: main.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+		log.Fatal(err)
+	}
+
+	_, err = f.WriteAt(metadataBytes, 0)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// getPendingMetadata fetches pending metadata from wpt.fyi/api/metadata/pending.
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply main.go:152]: Correct. It returns the same form but less number of metadata entries.

File: main.go [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+	testpath := path.Base(testName)
+	createDirIfNeeded(dirpath)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		log.Fatal(err)
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply main.go:45]: They are interchangeable in this case. I will replace `panic` with `log.Fatal`

File: main.go [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+func toWPTMetadata(testName string, link shared.MetadataLink) {
+	dirpath := path.Dir(testName)
+	testpath := path.Base(testName)
+	createDirIfNeeded(dirpath)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply main.go:43]: Good catch!

File: main.go [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{
+				{
+					Product: link.Product,
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply main.go:61]: Correct, there are optional `subtest` and `status` fields in `Results`

File: .github/workflows/test.yml (hunk 7 above)
Comment (KyleJu) [reply .github/workflows/test.yml:16]: Yea I will add test_main.go. This change is needed since we don't have to run `test_main.go` for every PR

File: main.go (hunk 11 above)
Comment (KyleJu) [reply main.go:109]: Correct

File: main.go (hunk 13 above)
Comment (KyleJu) [reply main.go:138]: I was able to improve the vet issue by upgrading `"gopkg.in/yaml.v2" to `"gopkg.in/yaml.v3"`. I will link this TODO to  web-platform-tests/wpt.fyi#1957

File: main.go (hunk 8 above)
Comment (KyleJu) [reply main.go]: Yea I simplified the code. The smallest `atoms` of metadata is [looped here](https://github.com/web-platform-tests/wpt-metadata/pull/1139/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R24). I will refactor and add tests

File: main.go (hunk 13 above)
Comment (foolip) [reply main.go]: Sounds good!

**Review by foolip — APPROVED** [review]
Thanks Kyle! Is this ready to go now, or are there tests you're going to add too?

**Comment by KyleJu** [comment]
> Thanks Kyle! Is this ready to go now, or are there tests you're going to add too?

Ready to go! Time to let it do work :)
